README.md
# Optika

<img src="https://raw.githubusercontent.com/phadej/optika/master/optika-300.png" align="right" height="100" />

> Optics: modular data access

[![Build Status](https://secure.travis-ci.org/phadej/optika.svg?branch=master)](http://travis-ci.org/phadej/optika)
[![NPM version](https://badge.fury.io/js/optika.svg)](http://badge.fury.io/js/optika)
[![Dependency Status](https://david-dm.org/phadej/optika.svg)](https://david-dm.org/phadej/optika)
[![devDependency Status](https://david-dm.org/phadej/optika/dev-status.svg)](https://david-dm.org/phadej/optika#info=devDependencies)

## Getting Started

Install the module with: `npm install optika`

## Synopsis

```javascript
var o = require("optika");

var value = {
  foo: [
    { bar: 2, baz: 3 },
    { bar: 4, baz: 5 },
  ]
};

// [2, 4]
o.key("foo").traversed().key("bar").arrayOf(value);

// { foo: [ { bar: 9, baz: 3 }, { bar: 11, baz: 5 } ] }
o.key("foo").traversed().key("bar").over(value, function (x) {
  return x + 7;
});
```

## Motivation

[Immutable.js](https://facebook.github.io/immutable-js/) is great!
But working with immutable containers & nested records is un-easy:

```javascript
return data.update("months", months =>
  months.map(month =>
    month.update("days", days =>
      days.map(day => injectAuxData(day, auxiliaryData))
    )
  )
);
```

The `updateIn` isn't powerful enough to make the drilling less boilerplaty
(and less error-prone).

If you are a Haskell programmer, you might know that *lenses* are a solution
to this problem:

```haskell
data_ & months . traversed . days . traversed %~ \day ->
  injectAuxData day auxData

-- or without operators:
over
  (months . traversed . days . traversed)
  (\day -> injectAuxData day auxData)
  data_
```

And with this library you can write similar code in JavaScript:

```javascript
o.imkey("months").traversed().imkey("days").traversed().over(data, day =>
  injectAuxData(day, auxData)
);
```

## Documentation

There is small amount of run-time validation.
For example, if you try to *set* over `Getter`, the exception will be thrown:
```javascript
o.key("foo").to(function (x) { return x[0]; }).set(value, 42);
// throws: Trying to run Traversal operation via Getter
```

### Operations

- `get(this: Getter<S,A>, value: S): A`

- `set(this: Traversal<S,T,A,B>, value: S, b: B): T`

- `over(this: Traversal<S,T,A,B>, value: S, f: A => B): T`

- `review(this: Prism<S,T,A,B>, value: B): T`
- `review(this: Iso<S,T,A,B>, value: B): T`

Construct value thru `Prism` or `Iso`.

- `affineView(this: Affine<S,T,A,B>, def: A): A

For operation working with `Fold`, see `firstOf`.

- `arrayOf(this: Fold<S,T,A,B>, value: S): Array<A>`

- `sumOf(this: Fold<S,T,number,B>, value: S): number`

### Constructors

- `lens(getter: S => A, setter: (S, B) => T): Lens<S,T,A,B>`

- `traversed(): Traversal<F<A>,F<B>,A,B>`

   Creates traversal for everything with `.map` and `.forEach`.

- `to(f: S => A): Getter<S,A>`

### Convenient optics

- `key(K: keyof (S & T)): Lens<S,T,S[K],T[K]>`

  Works for *plain-old-javascriot-objects*, i.e. POJOs :)

- `idx(i: number)): Lens<Array<A>,Array<A>,A,A>`

- `imkey(K: keyof (S & T)): Lens<Record<S>,Record<T>,S[K],T[K]>`

  Works with everyting supporting `.get` and `.set`, e.g.
  [Immutable](http://facebook.github.io/immutable-js/).

- `imidx(i: number)): Lens<List<A>,List<A>,A,A>`

  Works with everyting supporting `.get` and `.set`, e.g.
  [Immutable](http://facebook.github.io/immutable-js/).

  *Note:* doesn't perform bounds check.

- `filtered(pred: A => boolean): Prism<A,A,A',A'>`

*Note*: The predicate is checked when the value is injected (via `Traversal`)
or constructed via `Prism`.

- `safeFiltered(pred: A => boolean): Prism<A,A,A',A'>`

Like `filtered` but predicate is checked on construction.

### Internals

Functions which you probably never need to use directly.

`.o(this: Optic<S,T,A,B>, other: Optic<A,B,U,V>): Optic<S,T,U,V>`

Compose two optics.

`.run(this: Optic<S,T,A,B>, p: Profunctor<A,B>): Profunctor<S,T>`

## Contributing

- `README.md` is generated from the source with [ljs](https://github.com/phadej/ljs), say `make literate`.
- `optika.standalone.js` is also generated by the build process
- Before creating a pull request run `make test`, yet travis will do it for you.

## Release History

- **0.0.1** &mdash; *2017-03-24* &mdash; More features
- **0.0.0** &mdash; *2017-03-22* &mdash; Initial preview

## Implementation details

- pair is represented as two-element array: `Pair<A,B> = [A,B]`.
- `Either` is represtented by `[boolean, A | B]`, where
  `false` and `true` reprsent whether the value is `Left` or `Right` respectively.
- `Maybe` is encoded as the value + some unique value nothing can equal to (think: `Symbol`).

## Related work

### JavaScript

- [partial.lenses](https://github.com/calmm-js/partial.lenses)
- [ramda-lens](https://github.com/ramda/ramda-lens)
- [flunc-optics](https://github.com/flunc/optics)

The MIT License (MIT)

Copyright (c) 2017 Oleg Grenrus

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
## Release History

- **0.0.2** &mdash; *2017-04-01* &mdash; Neglect for speedups
- **0.0.1** &mdash; *2017-03-24* &mdash; More features
- **0.0.0** &mdash; *2017-03-22* &mdash; Initial preview


test/examples.js
/* jshint node:true */
/* global describe, it:true */
"use strict";

var o = require("../lib/optika.js");
var assert = require("assert");
var im = require("immutable");

var value = {
  foo: [
    { bar: 2, baz: 3 },
    { bar: 4, baz: 5 },
  ],
  quu: "foobar",
};

var imValue = im.fromJS(value);

function nonEmpty(x) { return x !== ""; }
function constTrue() { return true; }
function odd(x) { return x % 2 === 1; }
function inc2(x) { return x + 2; }

describe("examples", function () {
  describe("arrayOf", function () {
    it("readme", function () {
      var actual = o.key("foo").traversed().key("bar").arrayOf(value);
      var expected = [2, 4];
      assert.deepStrictEqual(actual, expected);
    });

    it("array", function () {
      var xs = [1, 2, 3, 4, 5];
      var actual = o.traversed().arrayOf(xs);
      var expected = xs;
      assert.deepStrictEqual(actual, expected);
    });

    it("immutable", function () {
      var actual = o.imkey("foo").traversed().imkey("bar").arrayOf(imValue);
      var expected = [2, 4];
      assert.deepStrictEqual(actual, expected);
    });

    it("with filtered", function () {
      var xss = [[1, 2], [3, 4, 5], [6, 7]];
      var actual = o.traversed().traversed().filtered(odd).arrayOf(xss);
      var expected = [1, 3, 5, 7];
      assert.deepStrictEqual(actual, expected);
    });
  });

  describe("forgetArrayOf", function () {
    it("readme", function () {
      var actual = o.key("foo").traversed().key("bar").forgetArrayOf(value);
      var expected = [2, 4];
      assert.deepStrictEqual(actual, expected);
    });

    it("array", function () {
      var xs = [1, 2, 3, 4, 5];
      var actual = o.traversed().forgetArrayOf(xs);
      var expected = xs;
      assert.deepStrictEqual(actual, expected);
    });

    it("immutable", function () {
      var actual = o.imkey("foo").traversed().imkey("bar").forgetArrayOf(imValue);
      var expected = [2, 4];
      assert.deepStrictEqual(actual, expected);
    });

    it("with filtered", function () {
      var xss = [[1, 2], [3, 4, 5], [6, 7]];
      var actual = o.traversed().traversed().filtered(odd).forgetArrayOf(xss);
      var expected = [1, 3, 5, 7];
      assert.deepStrictEqual(actual, expected);
    });
  });

  describe("sumOf", function () {
    it("readme", function () {
      var actual = o.key("foo").traversed().key("bar").sumOf(value);
      var expected = 6;
      assert.strictEqual(actual, expected);
    });

    it("array", function () {
      var xs = [1, 2, 3, 4, 5];
      var actual = o.traversed().sumOf(xs);
      var expected = 15;
      assert.strictEqual(actual, expected);
    });

    it("immutable", function () {
      var actual = o.imkey("foo").traversed().imkey("bar").sumOf(imValue);
      var expected = 6;
      assert.strictEqual(actual, expected);
    });

    it("with filtered", function () {
      var xss = [[1, 2], [3, 4, 5], [6, 7]];
      var actual = o.traversed().traversed().filtered(odd).sumOf(xss);
      var expected = 16;
      assert.strictEqual(actual, expected);
    });

    it("immutable with traversed at the end", function () {
      var xss = im.fromJS([[1, 2], [3, 4, 5], [6, 7]]);
      var actual = o.traversed().traversed().sumOf(xss);
      var expected = 28;
      assert.strictEqual(actual, expected);
    });
  });

  describe("forgetSumOf", function () {
    it("readme", function () {
      var actual = o.key("foo").traversed().key("bar").forgetSumOf(value);
      var expected = 6;
      assert.strictEqual(actual, expected);
    });

    it("array", function () {
      var xs = [1, 2, 3, 4, 5];
      var actual = o.traversed().forgetSumOf(xs);
      var expected = 15;
      assert.strictEqual(actual, expected);
    });

    it("immutable", function () {
      var actual = o.imkey("foo").traversed().imkey("bar").forgetSumOf(imValue);
      var expected = 6;
      assert.strictEqual(actual, expected);
    });

    it("with filtered", function () {
      var xss = [[1, 2], [3, 4, 5], [6, 7]];
      var actual = o.traversed().traversed().filtered(odd).forgetSumOf(xss);
      var expected = 16;
      assert.strictEqual(actual, expected);
    });

    it("immutable with traversed at the end", function () {
      var xss = im.fromJS([[1, 2], [3, 4, 5], [6, 7]]);
      var actual = o.traversed().traversed().forgetSumOf(xss);
      var expected = 28;
      assert.strictEqual(actual, expected);
    });
  });

  it("over", function () {
    var actual = o.key("foo").traversed().key("bar").over(value, function (x) {
      return x + 7;
    });
    var expected = {
      foo: [
        { bar: 9, baz: 3 },
        { bar: 11, baz: 5 },
      ],
      quu: "foobar",
    };
    assert.deepStrictEqual(actual, expected);
  });

  it("over: immutable", function () {
    var actual = o.imkey("foo").traversed().imkey("bar").over(imValue, function imm(x) {
      return x + 7;
    }).toJS();
    var expected = {
      foo: [
        { bar: 9, baz: 3 },
        { bar: 11, baz: 5 },
      ],
      quu: "foobar",
    };
    assert.deepStrictEqual(actual, expected);
  });

  it("over: with filtered", function () {
    var xss = [[1, 2], [3, 4, 5], [6, 7]];
    var actual = o.traversed().traversed().filtered(odd).over(xss, inc2);
    var expected = [[3, 2], [5, 4, 7], [6, 9]];
    assert.deepStrictEqual(actual, expected);
  });

  describe("set", function () {
    it("simple", function () {
      var actual = o.key("foo").traversed().key("bar").set(value, 42);
      var expected = {
        foo: [
          { bar: 42, baz: 3 },
          { bar: 42, baz: 5 },
        ],
        quu: "foobar",
      };
      assert.deepStrictEqual(actual, expected);
    });

    it("idx", function () {
      var actual = o.key("foo").idx(1).key("bar").set(value, 42);
      var expected = {
        foo: [
          { bar: 2, baz: 3 },
          { bar: 42, baz: 5 },
        ],
        quu: "foobar",
      };
      assert.deepStrictEqual(actual, expected);
    });

    it("imidx", function () {
      var actual = o.imkey("foo").imidx(1).imkey("bar").set(imValue, 42).toJS();
      var expected = {
        foo: [
          { bar: 2, baz: 3 },
          { bar: 42, baz: 5 },
        ],
        quu: "foobar",
      };
      assert.deepStrictEqual(actual, expected);
    });
  });

  describe("custom", function () {
    it("lens get", function () {
      var id = function (x) { return x; };
      var l = o.lens(id, id);

      var actual = l.get(42);
      var expected = 42;
      assert.strictEqual(actual, expected);
    });

    it("lens set", function () {
      var id = function (x) { return x; };
      var l = o.lens(id, function (s, b) { return b; });

      var actual = l.set(42, 7);
      var expected = 7;
      assert.strictEqual(actual, expected);
    });
  });

  describe("getter", function () {
    it("simple", function () {
      var actual = o.key("foo").to(function (xs) { return xs[0]; }).get(value);
      var expected = value.foo[0];
      assert.strictEqual(actual, expected);
    });

    it("idx", function () {
      var actual = o.key("foo").idx(1).key("bar").get(value);
      var expected = 4;
      assert.strictEqual(actual, expected);
    });

    it("imidx", function () {
      var actual = o.imkey("foo").imidx(1).imkey("bar").get(imValue);
      var expected = 4;
      assert.strictEqual(actual, expected);
    });
  });

  it("set with getter throws", function () {
    assert.throws(function () {
      o.key("foo").to(function (x) { return x[0]; }).set(value, 42);
    });
  });

  describe("affineView", function () {
    it("existing value", function () {
      var actual = o.key("quu").affineView(value, "none");
      var expected = "foobar";
      assert.strictEqual(actual, expected);
    });

    it("non-existing value", function () {
      var actual = o.key("quu").filtered(nonEmpty).affineView({ quu: "" }, "none");
      var actual2 = o.key("quu").safeFiltered(nonEmpty).affineView({ quu: "" }, "none");
      var expected = "none";
      assert.strictEqual(actual, expected);
      assert.strictEqual(actual2, expected);
    });

    it("nested non-existing", function () {
      var actual = o.key("quu").filtered(constTrue).filtered(nonEmpty).affineView({ quu: "" }, "none");
      var expected = "none";
      assert.strictEqual(actual, expected);
    });
  });

  describe("review", function () {
    it("works thru prism", function () {
      var actual = o.filtered(nonEmpty).review("foobar");
      var actual2 = o.safeFiltered(nonEmpty).review("foobar");
      var expected = "foobar";
      assert.strictEqual(actual, expected);
      assert.strictEqual(actual2, expected);
    });

    it("works thru violating filtered", function () {
      var actual = o.filtered(nonEmpty).review("");
      var expected = "";
      assert.strictEqual(actual, expected);
    });

    it("throws thru violation safeFiltered", function () {
      assert.throws(function () {
        o.safeFiltered(nonEmpty).review("");
      });
    });
  });

  describe("regressions", function () {
    it("axay", function () {
      var axay = [{ x: [{ y: 1 }] }];
      var actual = o.idx(0).key("x").idx(0).key("y").get(axay);
      var expected = 1;
      assert.strictEqual(actual, expected);
    });

    it("arrayOf doesn't mutate", function () {
      var ax = [1, 1, 1];
      var bx = im.fromJS([2, 2]);

      var actual = o.traversed().traversed().arrayOf([ax, [], bx]);
      var expected = [1, 1, 1, 2, 2];
      assert.deepStrictEqual(actual, expected);
      assert.deepStrictEqual(ax, [1, 1, 1]);
    });
  });

  // todo: firstOf
});


lib/utils.js
"use strict";

function isString(s) {
  return typeof s === "string";
}

function isNumber(n) {
  return typeof n === "number";
}

function isFunction(f) {
  return typeof f === "function";
}

function isArray(a) {
  return Array.isArray(a);
}

function assert(cond, msg) {
  if (!cond) {
    if (isFunction(msg)) {
      msg = msg();
    }

    throw new Error(msg);
  }
}

function identity(x) {
  return x;
}

function hasOwn(obj, prop) {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

/*
function fst(p) {
  return p[0];
}
*/

function snd(p) {
  return p[1];
}

function arraySingleton(x) {
  return [x];
}

module.exports = {
  arraySingleton: arraySingleton,
  assert: assert,
  hasOwn: hasOwn,
  identity: identity,
  isArray: isArray,
  isFunction: isFunction,
  isNumber: isNumber,
  isString: isString,
  snd: snd,
};


lib/monoid.js
"use strict";

var u = require("./utils.js");

var array = {
  init: function () { return []; },
  empty: [],
  snoc: function (xs, x) {
    xs.push(x);
    return xs;
  },
  append: function (a, b) {
    if (u.isArray(b)) {
      /* if (a.length === 0) {
        return a.concat(b); // we cannot return just `b`, as it might be mutated.
      } else */ if (b.length === 0) {
        return a;
      } else {
        return a.concat(b);
      }
    } else {
      b.forEach(function (x) {
        a.push(x);
      });
      return a;
    }
  },
  foldMap: function (f, xs) {
    var acc = [];
    if (u.isArray(xs)) {
      if (f === u.arraySingleton) {
        return xs;
      } else {
        var i = 0;
        var l = xs.length;
        while (i < l) {
          acc = acc.concat(f(xs[i]));
          i += 1;
        }
        return acc;
      }
    } else {
      xs.forEach(function (x) {
        acc = acc.concat(f(x));
      });
      return acc;
    }
  },
};

var sum = {
  init: function () { return 0; },
  empty: 0,
  snoc: function (a, b) {
    return a + b;
  },
  append: function (acc, xs) {
    if (u.isArray(xs)) {
      var i = 0;
      var l = xs.length;
      while (i < l) {
        acc += xs[i];
        i += 1;
      }

    } else {
      xs.forEach(function (x) {
        acc += x;
      });
    }

    return acc;
  },
  foldMap: function (f, xs) {
    var acc = 0;
    if (u.isArray(xs)) {
      var i = 0;
      var l = xs.length;
      while (i < l) {
        acc += f(xs[i]);
        i += 1;
      }
      return acc;
    } else {
      xs.forEach(function (x) {
        acc += f(x);
      });
      return acc;
    }
  },
};

module.exports = {
  array: array,
  sum: sum,
};


lib/convenience.js
"use strict";

var u = require("./utils.js");

function key(k, x0) {
  /* jshint validthis:true */
  u.assert(arguments.length === 2, "key: there should be 2 arguments");
  u.assert(u.isString(k), "key: k should be a string");

  // eslint-disable-next-line no-invalid-this
  var dict = this;

  function getter(s) {
    return [s[k], s];
  }

  function setter(p) {
    var s = p[1];
    var t = {};
    for (var l in s) {
      if (l === k) {
        t[l] = p[0];
      } else if (u.hasOwn(s, l)) {
        t[l] = s[l];
      }
    }
    return t;
  }

  // execution
  var x1 = dict.first(x0);
  var x2 = dict.dimap(getter, setter, x1);
  return x2;
}

// convenience adds convenience functions to the dictionaries
function convenience(dict) {
  if (dict.classes.indexOf("strong") !== -1) {
    if (!dict.key) { dict.key = key; }
  }

  return dict;
}

module.exports = convenience;


lib/optika.js
/**
  # Optika

  <img src="https://raw.githubusercontent.com/phadej/optika/master/optika-300.png" align="right" height="100" />

  > Optics: modular data access

  [![Build Status](https://secure.travis-ci.org/phadej/optika.svg?branch=master)](http://travis-ci.org/phadej/optika)
  [![NPM version](https://badge.fury.io/js/optika.svg)](http://badge.fury.io/js/optika)
  [![Dependency Status](https://david-dm.org/phadej/optika.svg)](https://david-dm.org/phadej/optika)
  [![devDependency Status](https://david-dm.org/phadej/optika/dev-status.svg)](https://david-dm.org/phadej/optika#info=devDependencies)

  ## Getting Started

  Install the module with: `npm install optika`

  ## Synopsis

  ```javascript
  var o = require("optika");

  var value = {
    foo: [
      { bar: 2, baz: 3 },
      { bar: 4, baz: 5 },
    ]
  };

  // [2, 4]
  o.key("foo").traversed().key("bar").arrayOf(value);

  // { foo: [ { bar: 9, baz: 3 }, { bar: 11, baz: 5 } ] }
  o.key("foo").traversed().key("bar").over(value, function (x) {
    return x + 7;
  });
  ```

  ## Motivation

  [Immutable.js](https://facebook.github.io/immutable-js/) is great!
  But working with immutable containers & nested records is un-easy:

  ```javascript
  return data.update("months", months =>
    months.map(month =>
      month.update("days", days =>
        days.map(day => injectAuxData(day, auxiliaryData))
      )
    )
  );
  ```

  The `updateIn` isn't powerful enough to make the drilling less boilerplaty
  (and less error-prone).

  If you are a Haskell programmer, you might know that *lenses* are a solution
  to this problem:

  ```haskell
  data_ & months . traversed . days . traversed %~ \day ->
    injectAuxData day auxData

  -- or without operators:
  over
    (months . traversed . days . traversed)
    (\day -> injectAuxData day auxData)
    data_
  ```

  And with this library you can write similar code in JavaScript:

  ```javascript
  o.imkey("months").traversed().imkey("days").traversed().over(data, day =>
    injectAuxData(day, auxData)
  );
  ```
*/

"use strict";

/**
  ## Documentation

  There is small amount of run-time validation.
  For example, if you try to *set* over `Getter`, the exception will be thrown:
  ```javascript
  o.key("foo").to(function (x) { return x[0]; }).set(value, 42);
  // throws: Trying to run Traversal operation via Getter
  ```
*/

// var assert = require("assert");

var monoid = require("./monoid.js");
var profunctor = require("./profunctor.js");
var u = require("./utils.js");

function Optic(classes, parts) {
  u.assert(u.isString(classes), "new Optic: classes should be string");
  u.assert(u.isArray(parts), "new Optic: parts should be an array");

  this.classes = classes;
  this.parts = parts;
}

/**
  ### Operations
*/

/**
  - `get(this: Getter<S,A>, value: S): A`
*/
Optic.prototype.get = function (value) {
  var f = this.run(profunctor.dictForgetNone, u.identity);
  var x = f(value);
  return x;
};

/**
  - `set(this: Traversal<S,T,A,B>, value: S, b: B): T`
*/
Optic.prototype.set = function (value, b) {
  return this.over(value, function () { return b; });
};

/**
  - `over(this: Traversal<S,T,A,B>, value: S, f: A => B): T`
*/
Optic.prototype.over = function (value, g) {
  var f = this.run(profunctor.dictFunc, g);
  var x = f(value);
  return x;
};

/**
  - `review(this: Prism<S,T,A,B>, value: B): T`
  - `review(this: Iso<S,T,A,B>, value: B): T`

  Construct value thru `Prism` or `Iso`.
*/
Optic.prototype.review = function (value) {
  return this.run(profunctor.dictTagged, value);
};

/**
  - `affineView(this: Affine<S,T,A,B>, def: A): A

  For operation working with `Fold`, see `firstOf`.
*/
Optic.prototype.affineView = function (value, def) {
  var f = this.run(profunctor.dictForgetMaybe, u.identity);
  var x = f(value);
  return x === profunctor.dictForgetMaybe.def ? def : x;
};

/**
  - `arrayOf(this: Fold<S,T,A,B>, value: S): Array<A>`
*/
Optic.prototype.arrayOf = function (value) {
  var f = this.run(profunctor.newDictForget(monoid.list), u.arraySingleton);
  var x = f(value);
  return x;
};

/**
   - `sumOf(this: Fold<S,T,number,B>, value: S): number`
*/
Optic.prototype.sumOf = function (value) {
  var f = this.run(profunctor.newDictForget(monoid.sum), u.identity);
  var x = f(value);
  return x;
};

/**
  ### Constructors
*/

function lensSetter(tuple) {
  return tuple[1](tuple[0]);
}

/**
 - `lens(getter: S => A, setter: (S, B) => T): Lens<S,T,A,B>`
*/
function lens(getter, setter) {
  u.assert(u.isFunction(getter), "lens: getter should be a function");
  u.assert(u.isFunction(setter), "lens: setter should be a function");

  function getter2(s) {
    return [getter(s), function (b) { return setter(s, b); }];
  }

  return new Optic("strong", [function (dict, x) {
    var x0 = dict.first(x);
    var x1 = dict.dimap(getter2, lensSetter, x0);
    return x1;
  }]);
}

/**
 - `traversed(): Traversal<F<A>,F<B>,A,B>`

    Creates traversal for everything with `.map` and `.forEach`.
*/
function traversed() {
  return new Optic("wander", [function (dict, x) {
    return dict.wander(x);
  }]);
}

/**
  - `to(f: S => A): Getter<S,A>`
*/
function to(f) {
  return new Optic("bicontra", [function (dict, x) {
    return dict.cimap(f, f, x);
  }]);
}

/**
  ### Convenient optics
*/

/**
  - `key(K: keyof (S & T)): Lens<S,T,S[K],T[K]>`

    Works for *plain-old-javascriot-objects*, i.e. POJOs :)
*/
function key(k) {
  u.assert(u.isString(k), "key: k should be a string");

  function getter(s) {
    return [s[k], s];
  }

  function setter(p) {
    var s = p[1];
    var t = {};
    for (var l in s) {
      if (l === k) {
        t[l] = p[0];
      } else if (u.hasOwn(s, l)) {
        t[l] = s[l];
      }
    }
    return t;
  }

  return new Optic("strong", [function (dict, x0) {
    var x1 = dict.first(x0);
    var x2 = dict.dimap(getter, setter, x1);
    return x2;
  }]);
}

/**
  - `idx(i: number)): Lens<Array<A>,Array<A>,A,A>`
*/
function idx(i) {
  u.assert(u.isNumber(i), "idx: i should be a number");

  function getter(s) {
    return [s[i], s];
  }

  function setter(p) {
    var t = p[1].slice();
    t[i] = p[0];
    return t;
  }

  return new Optic("strong", [function (dict, x0) {
    var x1 = dict.first(x0);
    var x2 = dict.dimap(getter, setter, x1);
    return x2;
  }]);
}

/**
  - `imkey(K: keyof (S & T)): Lens<Record<S>,Record<T>,S[K],T[K]>`

    Works with everyting supporting `.get` and `.set`, e.g.
    [Immutable](http://facebook.github.io/immutable-js/).
*/
function imkey(k) {
  u.assert(u.isString(k), "imkey: k should be a string");

  function getter(s) {
    return [s.get(k), s];
  }

  function setter(p) {
    return p[1].set(k, p[0]);
  }

  return new Optic("strong", [function (dict, x0) {
    var x1 = dict.first(x0);
    var x2 = dict.dimap(getter, setter, x1);
    return x2;
  }]);
}

/**
  - `imidx(i: number)): Lens<List<A>,List<A>,A,A>`

    Works with everyting supporting `.get` and `.set`, e.g.
    [Immutable](http://facebook.github.io/immutable-js/).

    *Note:* doesn't perform bounds check.
*/
function imidx(i) {
  u.assert(u.isNumber(i), "imidx: i should be a number");

  function getter(s) {
    return [s.get(i), s];
  }

  function setter(p) {
    return p[1].set(i, p[0]);
  }

  return new Optic("strong", [function (dict, x0) {
    var x1 = dict.first(x0);
    var x2 = dict.dimap(getter, setter, x1);
    return x2;
  }]);
}

// prism :: forall s t a b. (b -> t) -> (s -> Either t a) -> Prism s t a b
// prism to fro pab = dimap fro (either id id) (right (rmap to pab))

/**
  - `filtered(pred: A => boolean): Prism<A,A,A',A'>`

  *Note*: The predicate is checked when the value is injected (via `Traversal`)
  or constructed via `Prism`.
*/
function filtered(pred) {
  u.assert(u.isFunction(pred), "filtered: pred should be a function");

  // s -> Either t a
  function filteredGetter(x) {
    return [pred(x), x];
  }

  return new Optic("choice", [function (dict, x) {
    var x0 = dict.right(x);
    var x1 = dict.dimap(filteredGetter, u.snd, x0);
    return x1;
  }]);
}

/**
  - `safeFiltered(pred: A => boolean): Prism<A,A,A',A'>`

  Like `filtered` but predicate is checked on construction.
*/
function safeFiltered(pred) {
  u.assert(u.isFunction(pred), "safeFiltered: pred should be a function");

  // s -> Either t a
  function filteredGetter(x) {
    return [pred(x), x];
  }

  function check(x) {
    // TODO: throw own exception.
    u.assert(pred(x), "safeFiltered: construction of invalid value");
    return x;
  }

  return new Optic("choice", [function (dict, x) {
    var x0 = dict.dimap(u.identity, check, x);
    var x1 = dict.right(x0);
    var x2 = dict.dimap(filteredGetter, u.snd, x1);
    return x2;
  }]);
}

/**
  ### Internals

  Functions which you probably never need to use directly.
*/

/**
  `.o(this: Optic<S,T,A,B>, other: Optic<A,B,U,V>): Optic<S,T,U,V>`

  Compose two optics.
*/
Optic.prototype.o = function o(other) {
  var self = this;
  var classes = profunctor.normaliseClasses(self.classes, other.classes);

  return new Optic(classes, other.parts.concat(this.parts));
};

/**
  `.run(this: Optic<S,T,A,B>, p: Profunctor<A,B>): Profunctor<S,T>`
*/
Optic.prototype.run = function (dict, x) {
  var thisClasses = this.classes;
  var combined = profunctor.normaliseClasses(thisClasses, dict.classes);
  u.assert(combined === dict.classes, function () {
    return [
      "Trying to run",
      profunctor.classesToOpticName(dict.classes),
      "operation via",
      profunctor.classesToOpticName(thisClasses),
    ].join(" ");
  });

  var parts = this.parts;
  for (var i = 0; i < parts.length; i++) {
    x = parts[i](dict, x);
  }
  return x;
};

// Exports

var constructors = [
  filtered,
  idx,
  imidx,
  imkey,
  key,
  lens,
  safeFiltered,
  to,
  traversed,
];

var optics = {};
constructors.forEach(function (ctr) {
  // export
  optics[ctr.name] = ctr;

  // Optics.prototype postcomposed
  Optic.prototype[ctr.name] = function () {
    var other = ctr.apply(undefined, arguments);
    return this.o(other);
  };
});

// Additional exports

module.exports = optics;

// plain ../FAQ.md
/// plain ../CONTRIBUTING.md
/// plain ../CHANGELOG.md
/// plain ../implementation-details.md
/// plain ../related-work.md
/// plain ../LICENSE
